src/main/java/org/jenkinsci/plugins/workflow/cps/CpsThreadGroup.java

```diff
@@ -174,16 +174,13 @@ private Object readResolve() {
         execution = CpsFlowExecution.PROGRAM_STATE_SERIALIZATION.get();
         setupTransients();
         assert execution!=null;
-        if (scripts != null) { // compatibility: the field will be null in old programs
+        if (scripts != null && !scripts.isEmpty()) { // compatibility: the field will be null in old programs
             GroovyShell shell = execution.getShell();
-            assert shell.getContext().getVariables().isEmpty();
-            if (!scripts.isEmpty()) {
-                // Take the canonical bindings from the main script and relink that object with that of the shell and all other loaded scripts which kept the same bindings.
-                shell.getContext().getVariables().putAll(scripts.get(0).getBinding().getVariables());
-                for (Script script : scripts) {
-                    if (script.getBinding().getVariables().equals(shell.getContext().getVariables())) {
-                        script.setBinding(shell.getContext());
-                    }
+            // Take the canonical bindings from the main script and relink that object with that of the shell and all other loaded scripts which kept the same bindings.
+            shell.getContext().getVariables().putAll(scripts.get(0).getBinding().getVariables());
+            for (Script script : scripts) {
```

Comment:
And is the equality check on the next line still going to work under the circumstances you describe? The new test does not cover this case (e.g., use of `load` step).

Reply:
Good point, we could add a test like `RestartingLoadStepTest.accessToSiblingScripts` that also has variables injected via a `GroovyShellDecorator` to test the combined behavior.

Reply:
addressed in f15b02ddf3268eac1df80d25f14745975d68baa4

Reply:
2afbe31 removed the equality check and now applies all script bindings to the shell context.  @dwnusbaum created a test case called `updatedBindingsOnRestart` where the loaded script carries a stale definition if it is not given the shell context.

Reply:
To expand on this, we couldn't come up with a case where you didn't want the bindings to always be mapped to the same instance of `Binding`. Before the restart, at least in the tests cases we came up with, the main script and loaded scripts always shared the same instance of `Binding`. After the restart, the loaded scripts would have the same bindings as or a subset of the bindings of the main script (but never a superset), because some bindings that deserialize correctly as part of the main script get silently dropped in the binding for the the loaded script (I am not sure why, the ones that were dropped from the loaded script had values with type `CpsClosure` and other Groovy and CPS-specific types. Simple values seem to be ok.)

When the equality check was in place, if the main script did not have any bindings that the loaded scripts silently failed to deserialize, the check would pass, and the bindings would be linked. If the main script had a binding that could not be deserialized, as in the case with `accessToSiblingVariables`, the condition would pass for the main script but not the sibling scripts, which caused a bug after the restart as seen in `updatedBindingsOnRestart`. Since we couldn't come up with a case where it made sense for the equality check to fail (please let us know if you have any ideas), it seemed best to just remove it to fix `updatedBindingsOnRestart`.